lib/subscribers/langchain/base.js

```diff
@@ -0,0 +1,196 @@
+    this.recordCompletions({
+      events,
+      completionSummary,
+      segment,
+      transaction
+    })
+
+    if (err) {
+      agent.errors.add(
+        transaction,
+        err,
+        new LlmErrorMessage({
+          response: {},
+          cause: err,
+          summary: completionSummary
+        })
+      )
+    }
+
+    transaction.trace.attributes.addAttribute(DESTINATIONS.TRANS_EVENT, 'llm', true)
+  }
+
+  /**
+   * Records the LlmChatCompletionMessage(s)
+   *
+   * @param {object} params function params
+   * @param {Array} params.events prompt and response messages
+   * @param {LangChainCompletionSummary} params.completionSummary LlmChatCompletionSummary event
+   * @param {TraceSegment} params.segment active segment
+   * @param {Transaction} params.transaction active transaction
+   */
+  recordCompletions({ events, completionSummary, segment, transaction }) {
+    const { agent, logger } = this
+    for (let i = 0; i < events.length; i += 1) {
+      let msg = events[i]
+      if (msg?.content) {
+        msg = msg.content
+      }
+
+      let msgString
+      try {
+        msgString = typeof msg === 'string' ? msg : JSON.stringify(msg)
+      } catch (error) {
+        logger.error(error, 'Failed to stringify message')
+        msgString = ''
+      }
+
+      const completionMsg = new LangChainCompletionMessage({
+        sequence: i,
+        agent,
+        content: msgString,
+        completionId: completionSummary.id,
+        segment,
+        transaction,
+        runId: segment[langchainRunId],
+        // We call the final output in a LangChain "chain" the "response":
+        isResponse: i === events.length - 1
+      })
+
+      this.recordEvent({
+        type: 'LlmChatCompletionMessage',
+        pkgVersion: this.moduleVersion,
+        msg: completionMsg
+      })
+    }
+  }
+
+  /**
+   * Langchain allows you to define tags at the instance and call level
+   * This helper merges the two into 1 array ensuring there are not duplicates
+   *
+   * @param {Array} localTags tags defined on instance of a langchain object
+   * @param {Array} paramsTags tags defined on the method
+   * @returns {Array} a merged array of unique tags
+   */
+  mergeTags = function mergeTags(localTags = [], paramsTags = []) {
+    const tags = localTags.filter((tag) => !paramsTags.includes(tag))
+    tags.push(...paramsTags)
+    return tags
+  }
+
+  /**
+   * Langchain allows you to define metadata at the instance and call level
+   * This helper merges the two into object favoring the call level metadata
+   * values when duplicate keys exist.
+   *
+   * @param {object} localMeta metadata defined on instance of a langchain object
+   * @param {object} paramsMeta metadata defined on the method
+   * @returns {object} a merged object of metadata
+   */
+  mergeMetadata = function mergeMetadata(localMeta = {}, paramsMeta = {}) {
```

Comment:
this and mergeTags only apply to tools, shouldn't this just live in that subscriber?